compiler/include/preNormalizeOptimizations.h

```diff
@@ -24,11 +24,54 @@
 #include "CallExpr.h"
 #include "symbol.h"
 
+enum LocalityInfo { 
+  UNKNOWN,  // it may be anything
+  PENDING,  // we will make a decision later
+  LOCAL,    // we know this is local
+  UNAGGREGATABLE  // neither local, nor aggregatable
+};
+
+class AggregationCandidateInfo {
+  public:
+
+    CallExpr *candidate;
+    ForallStmt *forall;
+
+    LocalityInfo lhsLocalityInfo;
+    LocalityInfo rhsLocalityInfo;
+
+    CallExpr *lhsLogicalChild;
+    CallExpr *rhsLogicalChild;
+
+    // during normalize, we may generate aggregators for both sides of the
+    // assignment. However, after resolve we can have at most one aggregator per
+    // assignment ...
+    Symbol *srcAggregator;   // remote rhs
+    Symbol *dstAggregator;   // remote lhs
+
+    // ... and that's why we have a single aggregated alternative.
+    CallExpr *aggCall;
+
+    AggregationCandidateInfo(CallExpr *candidate, ForallStmt *forall);
+
+    void logicalChildAnalyzed(CallExpr *logicalChild, bool confirmed);
+    void addAggregators();
+  private:
+    void update();
+    void updateASTForAggregation(bool srcAggregation);
+    void updateASTForRegularAssignment();
+};
+
 // interface for normalize
 void doPreNormalizeArrayOptimizations();
 Symbol *earlyNormalizeForallIterand(CallExpr *call, ForallStmt *forall);
 
 // interface for resolution
 Expr *preFoldMaybeLocalThis(CallExpr *call);
+Expr *preFoldMaybeLocalArrElem(CallExpr *call);
+
+// interface for licm
+void transformConditionalAggregation(CondStmt *cond);
+void cleanupRemainingAggCondStmts();
```

Comment:
It seems strange that this file contains code not run before normalize (based on its name). Maybe you could rename it like prePostNormalizeOpts or something. Or move some of the code to separate files by optimization name. Or rename it "forallOptimization.h".

Reply:
@ronawho -- I remember you and I had a discussion around these lines. The name `forallOptimizations` sounds good to me, but I think you wanted to emphasize that some of this stuff needs to happen before normalization. Was that the case, or am I misremembering? If so, I could split the files somehow.

Reply:
I have a vague memory of that. https://github.com/chapel-lang/chapel/pull/16053#discussion_r453992122 is the only written conversation I found though.

Reply:
So, we either separate functions that are used for same/very similar optimizations into separate files, or we rename the file to have a more general name. I am currently leaning towards dropping "prenormalize" from the filename, as I feel like keeping things together is more important than specifying some detail of these optimizations in the filename. We have bunch of optimizations that happen after resolve, but we don't prefix them with `postResolve`, after all.

Michael also proposed:

> Or move some of the code to separate files by optimization name

Which was something I thought about but decided not to. Because, implementation-wise, all three optimizations are too tightly related. We can think of having a `forallAnalysis` and separate headers for optimizations that use functionality in there. But that's probably more than what we need right now.

Reply:
Of the options in front of us, I like renaming it to `forallOptimizations` the best. (We have other forall optimizations not in the file and I think that is OK).

Reply:
I renamed the file to `forallOptimizations`